File: app/src/main/java/com/tsaysoft/nfpachemidv3/IDGManager.java
package com.tsaysoft.nfpachemidv3;

import java.util.*;

/**
 * <code>IDGManager</code>s manage multiple {@link IDGAbstract}s.
 * <p>
 *     Past experience has shown that it is difficult to query chemical IDs
 *     from chemical names as many names have added terms (e.g. "inhibited")
 *     that interfere with name-to-ID conversion services. Therefore, <code>IDGManager</code>s
 *     take in information from numerous sources to arrive at an accurate chemical ID.
 * </p>
 *
 * @author Clay Tsay
 * @version 00.01.00
 */
public class IDGManager implements IDGInterface{

    // --------------------
    // VARIABLES AND DATA
    // --------------------

    private List<IDGAbstract> IDGs = new ArrayList<>();



    // --------------------
    // CONSTRUCTORS
    // --------------------

    /**
     * Constructs an empty <code>IDGManager</code>.
     *
     * @since 00.01.00
     */
    public IDGManager() {
        super();
    }

    /**
     * Constructs an <code>IDGManager</code> with one <code>IDGetter</code>.
     * <p>
     *     Does not possess the ability to determine between <code>null</code> and non-<code>null</code> inputs.
     * </p>
     *
     * @param IDG the <code>IDGetter</code> to be put into the <code>IDGManager</code>
     *
     * @since 00.01.00
     */
    public IDGManager(IDGAbstract IDG) {
        super();
        IDGs.add(IDG);
    }

    /**
     * Constructs an <code>IDGManager</code> with multiple <code>IDGetter</code>s.
     * <p>
     *     Does not possess the ability to determine between <code>null</code> and non-<code>null</code> inputs.
     * </p>
     *
     * @param IDGCollection a <code>Collection</code> of <code>IDGetter</code>s to be put into the <code>IDGManager</code>
     *
     * @since 00.01.00
     */
    public IDGManager(Collection<IDGAbstract> IDGCollection) {
        super();
        IDGs.addAll(IDGCollection);
    }

    // --------------------
    // PUBLIC UTILITY METHODS
    // --------------------

    @Override
    public String requestID(String chemName, ChemID id) {
        // Initialize variables
        List<String> results = new ArrayList<>();
        Map<String, Integer> idFreqs = new HashMap<>();
        Integer tempInt;
        String maxToken = null;
        int maxFreq = 0;

        // Ask all of the IDGetters for their IDs
        for(IDGAbstract i : IDGs) {
            results.add(i.requestID(chemName, id));
        }

        // Find the numerical frequencies of the IDs
        for(String token : results) {
            if(!idFreqs.containsKey(token)) {
                idFreqs.put(token, 1);
            } else {
                tempInt = idFreqs.get(token);
                idFreqs.put(token, tempInt+1);
            }
        }

        // Find the ID with the greatest frequency
        for(String key : idFreqs.keySet()) {
            if(idFreqs.get(key) > maxFreq) {
                maxFreq = idFreqs.get(key);
                maxToken = key;
            }
        }

        // If there is an ID with greatest frequency, return. If not, randomly choose.

        // TODO: Get this method working
        return maxToken;



    }

    // --------------------
    // PRIVATE UTILITY METHODS
    // --------------------

    private static String determineMax(List<String> strings) {
        //Initialise variables
        Map<String, Integer> map = new HashMap<String, Integer>();

        // Initialise frequency table
        for(String s : strings) {
            Integer freq = map.get(s);
            map.put(s, (freq == null) ? 1 : freq + 1);
        }
        return null;

    }


}

File: app/src/main/java/com/tsaysoft/nfpachemidv3/IDGManager.kt
package com.tsaysoft.nfpachemidv3

import java.util.*

/**
 * [IDGManager]s manage multiple [IDGAbstract]s.
 *
 * Past experience has shown that it is difficult to query chemical IDs
 * from chemical names as many names have added terms (e.g. "inhibited")
 * that interfere with name-to-ID conversion services. Therefore, [IDGManager]s
 * take in information from numerous sources to arrive at an accurate chemical ID.
 *
 * @author Clay Tsay
 * @version 00.01.00
 */
class IDGManager : IDGInterface {

    // --------------------
    // VARIABLES AND DATA
    // --------------------

    private val IDGs = ArrayList<IDGAbstract>()


    // --------------------
    // CONSTRUCTORS
    // --------------------

    /**
     * Constructs an empty `IDGManager`.
     *
     * @since 00.01.00
     */
    constructor() : super() {}

    /**
     * Constructs an [IDGManager] with one `IDGetter`.
     *
     * Does not possess the ability to determine between `null` and non-`null`
     * inputs.
     *
     * @param IDG the `IDGetter` to be put into the [IDGManager]
     * @since 00.01.00
     */
    constructor(IDG: IDGAbstract) : super() {
        IDGs.add(IDG)
    }

    /**
     * Constructs an [IDGManager] with multiple `IDGetter`s.
     *
     *
     * Does not possess the ability to determine between `null` and non-`null`
     * inputs.
     *
     * @param IDGCollection a [Collection] of `IDGetter`s to be put into
     *      the [IDGManager]
     * @since 00.01.00
     */
    constructor(IDGCollection: Collection<IDGAbstract>) : super() {
        IDGs.addAll(IDGCollection)
    }

    // --------------------
    // PUBLIC UTILITY METHODS
    // --------------------

    override fun requestID(chemName: String, id: ChemID): String {
        // Initialize variables
        val results = ArrayList<String>()
        val idFreqs = HashMap<String, Int>()
        var tempInt: Int?
        var maxToken: String? = null
        var maxFreq = 0

        // Ask all of the IDGetters for their IDs
        for (i in IDGs) {
            results.add(i.requestID(chemName, id))
        }

        // Find the numerical frequencies of the IDs
        // TODO: (somehow) null-proof this section of the programme
        for (token in results) {
            if (!idFreqs.containsKey(token)) {
                idFreqs[token] = 1
            } else {
                tempInt = idFreqs[token]
                idFreqs[token] = tempInt!! + 1
            }
        }

        // Find the ID with the greatest frequency
        // Maybe make it so that if there is a tie the ID with the greatest frequency is chosen.
        for (key in idFreqs.keys) {
            if (idFreqs[key] > maxFreq) {
                maxFreq = idFreqs[key]
                maxToken = key
            }
        }

        return maxToken

    }

    // --------------------
    // PRIVATE UTILITY METHODS
    // --------------------

    // TODO: Get this method working
    // You might also want to figure out what it's supposed to do
    private fun determineMax(strings: List<String>): String? {
        //Initialise variables
        val map = HashMap<String, Int>()

        // Initialise frequency table
        for (s in strings) {
            val freq = map[s]
            map[s] = if (freq == null) 1 else freq + 1
        }
        return null

    }


}
